File: Code/Chapter 5/C#/Example 5.4/Form1.cs
#region Using directives

using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Windows.Forms;

using System.Data.SqlClient ;
using System.Data.Common;
using System.Collections;

#endregion

namespace Example_5_4
{
    partial class Form1 : Form
    {
        private ArrayList dbRecordsHolder;
        public Form1()
        {
            InitializeComponent();
            dbRecordsHolder = new ArrayList();
        }

        private void btnPopulate_Click(object sender, EventArgs e)
        {
            string connectionString = "Server=(local);Database=Test;Integrated Security=SSPI;Asynchronous Processing=true";
            SqlConnection testConnection = new SqlConnection(connectionString);
            
            SqlCommand testCommand =
                new SqlCommand("Select * from TestDemo", testConnection);
            testConnection.Open();

            AsyncCallback callback = new AsyncCallback(DataReaderIsReady);
            IAsyncResult asyncresult = testCommand.BeginExecuteReader(callback, testCommand);
        }

        private void DataReaderIsReady(IAsyncResult result)
        {
            MessageBox.Show("Results Load Complete","I'm Done");
            SqlCommand testCommand = (SqlCommand)result.AsyncState;

            SqlDataReader sqlDr = testCommand.EndExecuteReader(result);
            if (sqlDr.HasRows)
            {
                foreach (DbDataRecord rec in sqlDr)
                {
                    dbRecordsHolder.Add(rec);
                }
            }
            sqlDr.Close();
            testCommand.Connection.Dispose();
        }

        private void btnDataBind_Click(object sender, EventArgs e)
        {
            myDataGrid.DataSource = dbRecordsHolder;
        }
    }
}

File: Code/Chapter 5/VB.NET/Exercise 5.4/Form1.vb
Imports System.Data
Imports System.Data.Common
Imports System.Collections
Imports System.Data.SqlClient

Public Class Form1

   Dim dbRecordsHolder As ArrayList

   Private Sub btnPopulate_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btnPopulate.Click
      Dim connectionString As String = _
         "Server=(local);Database=Test;Integrated Security=SSPI;Asynchronous Processing=true"
      Dim testConnection As SqlConnection = _
         New SqlConnection(connectionString)

      Dim testCommand As SqlCommand = _
         New SqlCommand("Select * from TestDemo", testConnection)
      testConnection.Open()

      Dim callback As AsyncCallback = New AsyncCallback(AddressOf DataReaderIsReady)
      Dim asyncresult As IAsyncResult = testCommand.BeginExecuteReader(callback, testCommand)
   End Sub

   Private Sub DataReaderIsReady(ByVal result As IAsyncResult)
      Dim testCommand As SqlCommand = CType(result.AsyncState, SqlCommand)

      Dim sqlDr As SqlDataReader = testCommand.EndExecuteReader(result)
      If sqlDr.HasRows Then

         Dim rec As DbDataRecord
         For Each rec In sqlDr
            dbRecordsHolder.Add(rec)
         Next
      End If
      sqlDr.Close()
      testCommand.Connection.Dispose()
   End Sub

   Private Sub btnDataBind_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btnDataBind.Click
      myDataGrid.DataSource = dbRecordsHolder
   End Sub

   Private Sub Form1_Load(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles MyBase.Load
      dbRecordsHolder = New ArrayList()
   End Sub
End Class
